adapters/beachfront/beachfront.go

```diff
@@ -197,289 +184,327 @@ func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *a
-			(request.Imp[i].Banner.H != nil && request.Imp[i].Banner.W != nil)) {
-			bannerImps = append(bannerImps, request.Imp[i])
+		if !gotBanner {
+			gotBanner = testForValidBanner(request, i)
 		}
 
 		if request.Imp[i].Video != nil {
-			videoImps = append(videoImps, request.Imp[i])
+			imp := request.Imp[i]
+			var secure int8 = 0
+
+			if request.Site != nil && request.Site.Page != "" {
+				secure = isPageSecure(request.Site.Page)
+				imp.Secure = &secure
+			} else {
+				// &secure = 0 at this point
+				if imp.Secure != nil && imp.Secure != &secure {
+					imp.Secure = request.Imp[i].Secure
+				} else {
+					imp.Secure = &secure
+				}
+			}
+
+			imp.Banner = nil
+			videoImps = append(videoImps, imp)
 		}
 	}
 
-	if len(bannerImps)+len(videoImps) == 0 {
-		errs = append(errs, errors.New("no valid impressions were found in the request"))
+	if len(videoImps) == 0 && !gotBanner {
+		*errs = append(*errs, errors.New("no valid impressions were found in the request"))
 		return
 	}
 
-	if len(bannerImps) > 0 {
-		request.Imp = bannerImps
-		beachfrontReqs.Banner, errs = getBannerRequest(request)
+	if gotBanner {
+		beachfrontReqs.Banner = getBannerRequest(request, errs)
 	}
 
 	if len(videoImps) > 0 {
-		var videoErrs []error
-		var videoList []beachfrontVideoRequest
-
-		request.Imp = videoImps
-
-		videoList, videoErrs = getVideoRequests(request)
-		errs = append(errs, videoErrs...)
+		requestStub := *request
+		requestStub.Imp = nil
 
-		for i := 0; i < len(videoList); i++ {
-			if videoList[i].VideoResponseType == "nurl" || videoList[i].VideoResponseType == "both" {
-				beachfrontReqs.NurlVideo = append(beachfrontReqs.NurlVideo, videoList[i])
-			}
-
-			if videoList[i].VideoResponseType == "adm" || videoList[i].VideoResponseType == "both" {
-				beachfrontReqs.ADMVideo = append(beachfrontReqs.ADMVideo, videoList[i])
-			}
-		}
+		beachfrontReqs.NurlVideo, beachfrontReqs.ADMVideo = getVideoRequests(requestStub, videoImps, errs)
 	}
-
 	return
 }
 
+/**
+Returns an authoritative appId (exchange_id) for a specific impression
+*/
 func getAppId(ext openrtb_ext.ExtImpBeachfront, media openrtb_ext.BidType) (string, error) {
 	var appid string
-	var error error
-
-	if fmt.Sprintf("%s", reflect.TypeOf(ext.AppId)) == "string" &&
-		ext.AppId != "" {
+	var err error
 
+	if ext.AppId != "" {
 		appid = ext.AppId
-	} else if fmt.Sprintf("%s", reflect.TypeOf(ext.AppIds)) == "openrtb_ext.ExtImpBeachfrontAppIds" {
+	} else {
 		if media == openrtb_ext.BidTypeVideo && ext.AppIds.Video != "" {
 			appid = ext.AppIds.Video
 		} else if media == openrtb_ext.BidTypeBanner && ext.AppIds.Banner != "" {
 			appid = ext.AppIds.Banner
 		}
-	} else {
-		error = errors.New("unable to determine the appId(s) from the supplied extension")
 	}
 
-	return appid, error
+	if appid == "" {
+		err = errors.New("unable to determine the appId(s) from the supplied extension")
+	}
+
+	if appid == "throwTestError" {
```

Comment:
Whats the rationale behind having a test specific app ID in the code here? Anything test specific should be in the test files. This isn't considered as a best practice

Reply:
Thank you. I'll find a better way of doing this. I need to dig into how the test files work a lot more thoroughly.